pyangbind/lib/serialise.py

```diff
@@ -164,54 +160,271 @@ def default(self, obj, mode="default"):
             return six.text_type(obj)
         elif map_val in ["pyangbind.lib.yangtypes.YANGBool"]:
-            if original_yang_type == "empty" and mode == "ietf":
-                if obj:
-                    return [None]
+            if original_yang_type == "empty":
+                # NOTE: previously with IETF mode the code would fall-through if obj was falsey
+                return self.yangt_empty(obj)
             else:
-                if obj:
-                    return True
-                else:
-                    return False
+                return bool(obj)
         elif map_val in ["pyangbind.lib.yangtypes.TypedList"]:
-            return [self.default(i) for i in obj]
-        elif map_val in ["int"]:
-            int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            return int(obj)
-        elif map_val in ["long"]:
+            return self.yangt_typed_list(obj)
+        elif map_val in ["int", "long"]:
             int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            elif six.PY3:
-                return int(obj)
-            else:
-                return long(obj)
+            return self.yangt_long(obj) if int_size == 64 else self.yangt_int(obj)
         elif map_val in ["container"]:
-            return self._preprocess_element(obj.get(), mode=mode)
+            return self.preprocess_element(obj.get())
         elif map_val in ["decimal.Decimal"]:
-            return six.text_type(obj) if mode == "ietf" else float(obj)
+            return self.yangt_decimal(obj)
+
+    def yangt_int(self, obj):
+        # for values that are 32-bits and under..
+        return int(obj)
+
+    def yangt_long(self, obj):
+        # don't need to sepcial-case for PY3 because we assign `long = int` in the header
+        return long(obj)
+
+    def yangt_identityref(self, obj):
+        raise UnmappedItem
+
+    def yangt_decimal(self, obj):
+        # doesn't seem to need a distinction for non-64-bit decimal?
+        return float(obj)
+
+    def yangt_empty(self, obj):
+        return bool(obj)
+
+    def yangt_typed_list(self, obj):
+        return [self.default(i) for i in obj]
+
+
+class IETFYangDataSerialiser(YangDataSerialiser):
+    """
+    IETF data serialiser overrides some of the data type formats only
+    """
+
+    def yangt_long(self, obj):
+        return six.text_type(obj)
+
+    def yangt_identityref(self, obj):
+        try:
+            emod = obj._enumeration_dict[obj]["@module"]
+            if emod != obj._defining_module:
+                return "%s:%s" % (obj._enumeration_dict[obj]["@module"], obj)
+        except KeyError:
+            pass
+        return six.text_type(obj)
+
+    def yangt_decimal(self, obj):
+        return six.text_type(obj)
+
+    def yangt_empty(self, obj):
```

Comment:
This isn't quite correct AFAICS. In general, we expect that an `empty` must be true or not set, if the value is False, we should not include the leaf in the serialised output.

Reply:
Hi Rob, to be honest this is quite new for me. You'll need to elaborate a bit more on what the desired behaviour should be. For now, I have basically tried to conform to the existing (previous) implementation which was this:

```python
         elif orig_yangt in ["empty"]:
            if obj:
                return [None] if mode == "ietf" else True
            return False
```

So my reading of this is basically the following truth table:

obj | IETF-JSON | vanilla JSON
--- | --- | ---
truthy | `[None]` | `True`
falsy | `False` | `False`

So I have the following implementation:

IETF-JSON | vanilla JSON
--- | ---
`[None] if obj else False` | `bool(obj)`

_However_, there is a note later in the code where this diverges from the existing implementation....

```python
       elif map_val in ["pyangbind.lib.yangtypes.YANGBool"]:
            if original_yang_type == "empty" and mode == "ietf":
                if obj:
                    return [None]
            else:
                if obj:
                    return True
                else:
                    return False
```

For an IETF empty with `obj` as false then it would fall through this clause. In this proposed PR this _no longer happens_! I *think* this is safe, but I have no idea about the greater context. It appears to work :)

Reply:
The issue is `empty` as a YANG type doesn't really have a bool representation. It's either present or not. In the generated classes, the choice was to represent it as a bool and only allow it to be set to true. So before, when the serialisation happened then we would fall through and cause a serialisation error.

Right now, the implementation doesn't prevent someone setting an empty to `False` at set-time, so this seemed a reasonable compromise. I think we should keep this as an error in the serialisation code either way.

For background of why `empty` works like this, it's really an XML element that is:

`<leaf />`

Clearly, this thing is either there (value nothing, it's there) or not.